language-plutus-core/test/Evaluation/TypeCheck.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Evaluation.TypeCheck
+    ( test_typecheck
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.StdLib.Data.Bool
+import           Language.PlutusCore.StdLib.Data.ChurchNat
+import           Language.PlutusCore.StdLib.Data.Function
+import           Language.PlutusCore.StdLib.Data.List
+import           Language.PlutusCore.StdLib.Data.Nat
+import           Language.PlutusCore.StdLib.Data.Unit
+
+import           Data.Foldable
+import qualified Data.Text.Encoding as Text
+import qualified Data.ByteString.Lazy as Bsl
+import           Test.Tasty
+import           Test.Tasty.HUnit
+
+-- | Assert a 'Term' is well-typed.
+assertQuoteWellTyped :: HasCallStack => Quote (Term TyName Name ()) -> Assertion
+assertQuoteWellTyped getTerm =
+    let term = runQuote getTerm in
+        for_ (typecheckTerm term) $ \err -> assertFailure $ concat
+            [ "Ill-typed: ", prettyCfgString term, "\n"
+            , "Due to: ", prettyCfgString err
+            ]
+
+-- | Assert a term is ill-typed.
+assertQuoteIllTyped :: HasCallStack => Quote (Term TyName Name ()) -> Assertion
+assertQuoteIllTyped getTerm =
+    let term = runQuote getTerm in
+        case typecheckTerm term of
+            Nothing -> assertFailure $ "Well-typed: " ++ prettyCfgString term
+            Just _  -> return ()
+
+typecheckProgram :: Program TyName Name () -> Maybe Error
```

Comment:
You might want to use the functions I recently added to `Quote.hs` for this kind of thing.